src/storage/txn/actions/check_txn_status.rs

```diff
@@ -5,16 +5,114 @@ use tikv_kv::SnapshotExt;
 use txn_types::{Key, Lock, TimeStamp, Write, WriteType};
 
 use crate::storage::{
+    mvcc,
     mvcc::{
         metrics::MVCC_CHECK_TXN_STATUS_COUNTER_VEC, reader::OverlappedWrite, ErrorInner, LockType,
         MvccTxn, ReleasedLock, Result, SnapshotReader, TxnCommitRecord,
     },
     Snapshot, TxnStatus,
 };
 
-// Check whether there's an overlapped write record, and then perform rollback.
-// The actual behavior to do the rollback differs according to whether there's
-// an overlapped write record.
+fn check_txn_status_from_pessimistic_primary_lock(
+    txn: &mut MvccTxn,
+    reader: &mut SnapshotReader<impl Snapshot>,
+    primary_key: Key,
+    lock: &Lock,
+    current_ts: TimeStamp,
+    resolving_pessimistic_lock: bool,
+) -> Result<(TxnStatus, Option<ReleasedLock>)> {
+    assert!(lock.is_pessimistic_lock());
+    // Check the storage information first in case the force lock could be stale.
+    // See https://github.com/pingcap/tidb/issues/43540 for more details.
+    if lock.is_pessimistic_lock_with_conflict() {
+        let (txn_status, is_current_lock_stale) = match check_txn_status_missing_lock(
+            txn,
+            reader,
+            primary_key.clone(),
+            None,
+            MissingLockAction::ReturnError,
```

Comment:
Sorry I didn't get it. In case the transaction record is not found from write cf, shouldn't we follow the `rollback_if_not_exist` parameter like how a normal invocation of `check_txn_status_missing_lock` does?

Reply:
If `is_pessimistic_lock_with_conflict` is true the pessimistic primary lock could be stale if there exists a commit or rollback record in the write cf. While if there's no corresponding transaction information in the write cf, the pessimistic primary lock is regarded as **valid**, and the transaction status checking continues as the normal path in `check_txn_status_lock_exists`. 
So `check_txn_status_missing_lock` is just used to check if there exists a corresponding record.

Reply:
Can we try to assert that `check_txn_status_missing_lock` called here won't perform any write operation?
It seems it won't write anything (to the `txn`) currently. I'm kind of afraid that if in the future we mistakenly writes the samething  (the same physical key) more than once in a single `MvccTxn` object, the final result might be undefined

Reply:
A new read-only method `check_txn_status_from_storage` is added and the `check_txn_status_missing_lock` is not reused to avoid possible mis-usages.